oasis-reference/src/main/kotlin/com/empowerops/volition/ref_oasis/optimizer/OptimizerEndPoint.kt

```diff
@@ -0,0 +1,151 @@
+            responseObserver: StreamObserver<NodeChangeConfirmDTO>
+    ) = responseObserver.consume {
+        checkThenRun(modelService.simulations.hasName(request.name)) {
+            apiService.autoConfigure(request)
+        }
+    }
+
+    override fun updateConfiguration(
+            request: ConfigurationCommandDTO,
+            responseObserver: StreamObserver<ConfigurationResponseDTO>
+    ) = responseObserver.consume {
+        checkThenRun(modelService.simulations.hasName(request.name)) {
+            apiService.updateConfiguration(request)
+        }
+    }
+
+    override fun requestRunResult(
+            request: ResultRequestDTO,
+            responseObserver: StreamObserver<ResultResponseDTO>
+    ) = responseObserver.consume {
+        checkThenRun(modelService.simulations.hasName(request.name)) {
+            apiService.resultRequest(request)
+        }
+    }
+
+    override fun registerRequest(
+            request: RequestRegistrationCommandDTO,
+            responseObserver: StreamObserver<RequestQueryDTO>
+    ) {
+            apiService.register(request, responseObserver)
+    }
+
+    override fun startOptimization(
+            request: StartOptimizationCommandDTO,
+            responseObserver: StreamObserver<StartOptimizationResponseDTO>
+    ) = responseObserver.consumeThen({
+        checkThenRun(modelService.simulations.hasName(request.name)) {
+            apiService.requestStart(request)
+        }
+    }) { response ->
+        if (response.acknowledged) apiService.start()
+    }
+
+    override fun stopOptimization(
+            request: StopOptimizationCommandDTO,
+            responseObserver: StreamObserver<StopOptimizationResponseDTO>
+    ) = responseObserver.consumeThen({
+        checkThenRun(modelService.simulations.hasName(request.name)) {
+            apiService.requestStop(request)
+        }
+    }) { response ->
+        if (response.responseCase == StopOptimizationResponseDTO.ResponseCase.RUNID) apiService.stop()
+    }
+
+    override fun updateNode(
+            request: NodeStatusCommandOrResponseDTO,
+            responseObserver: StreamObserver<NodeChangeConfirmDTO>
+    ) = responseObserver.consume {
+        checkThenRun(modelService.simulations.hasName(request.name)) {
+            apiService.updateNode(request)
+        }
+    }
+
+    override fun sendMessage(
+            request: MessageCommandDTO,
+            responseObserver: StreamObserver<MessageResponseDTO>
+    ) = responseObserver.consume {
+        checkThenRun(modelService.simulations.hasName(request.name)) {
+            apiService.sendMessage(request)
+        }
+    }
+
+    override fun unregisterRequest(
+            request: RequestUnRegistrationRequestDTO,
+            responseObserver: StreamObserver<UnRegistrationResponseDTO>
+    ) = responseObserver.consume {
+        checkThenRun(modelService.simulations.hasName(request.name)) {
+            apiService.unregister(request)
+        }
```

Comment:
oh but i see, this has become something of a pattern for you. The Optimizer Endpoint actually does the state validation. 

So the optimizer endpoint is responsible for checking that the operation is legal, then asking the API service to perform that operation. If its not legal, you've written some idioms to have that project into a `RequestFailedDTO`. 

how is `CheckThenRun` able to return both `MessageResponseDTO`s and `UnRegistrationReponseDTO`s?

Reply:
Sound interesting, I will take a look. The other reason we got here is also me thinking sequentially: we should tell caller we are starting then send here is the run that started but I think if we manage the asynchronous better we shouldn't be worrying about that. 

One thing I was just trying to use `consume` we had
it could just be 
```Kotlin
fun(reqeust: Type, responseObserver: ObserverType){
  //do check and build response
  responseObserver.consume(response)
  //do the rest
}
```